docs: First pass at extensions documentation (#750)
google-a2a/A2A: docs/topics/extensions.md

# Extensions

## Abstract

> Extensions are a means of extending the Agent2Agent (A2A) protocol with new data, requirements, methods, and state machines. Agents declare their support for extensions in their AgentCard, and clients can then opt-in to the behavior offered by the extension as part of requests they make to the agent. Extensions are identified by a URI and defined by their extension specification. Anyone is able to define, publish, and implement an extension.

Reviewer: ```suggestion
Extensions are a means of extending the Agent2Agent (A2A) protocol with new data, requirements, methods, and state machines. Agents declare their support for extensions in their [`AgentCard`](/specification/#5-agent-discovery-the-agent-card), and clients can then opt-in to the behavior offered by the extension as part of requests they make to the agent. Extensions are identified by a URI and defined by their extension specification. Anyone is able to define, publish, and implement an extension.
```

## Introduction

The core A2A protocol is a solid basis for enabling communication between agents. However, it's clear that some domains require additional structure than what is offered by the generic methods in the protocol. Extensions were added to the protocol to help support these cases: with extensions, agents and clients can negotiate additional, custom logic to be layered on top of the core protocol.

### Scope of Extensions

The exact set of possible ways to use extensions is intentionally not defined. This is to facilitate the ability to use extensions to expand A2A beyond currently known use cases. However, some use cases are clearly forseeable, such as:

> - Exposing new information in the AgentCard. An extension may not impact the request/response flow at all -- it can be simply used as a way to convey additional structured information to clients via the AgentCard. We refer to these as *data-only extensions*.

Reviewer: ```suggestion
- Exposing new information in the `AgentCard`. An extension may not impact the request/response flow at all -- it can be simply used as a way to convey additional structured information to clients via the `AgentCard`. We refer to these as *data-only extensions*.
```

> - Overlaying additional structure and state change requirements on the core request/response messages. An extension could, for example, require that all messages use DataParts that adhere to a specific schema. This type of extension effectively acts as a profile on the core A2A protocol, narrowing the space of allowed values. We refer to these as *profile extensions*.

Reviewer: ```suggestion
- Overlaying additional structure and state change requirements on the core request/response messages. An extension could, for example, require that all messages use `DataPart`s that adhere to a specific schema. This type of extension effectively acts as a profile on the core A2A protocol, narrowing the space of allowed values. We refer to these as *profile extensions*.
```

- Adding new RPC methods entirely. Extensions may define that the agent implements more than the core set of protocol methods. We refer to these as *method extensions*.

There are some changes to the protocol that extensions *do not* allow. These are:

- Changing the definition of core data structures. Adding new fields or removing required fields to protocol-defined data structures is not supported. Extensions are expected to place custom attributes in the `metadata` map that is present on core data structures.
- Adding new values to enum types. Instead, extensions should use existing enum values and annotate additional semantic meaning in the `metadata` field.

These limitations exist to prevent extensions from breaking core type validations that clients and agents perform.

### Architecture Overview

> [TODO: Diagram]

Reviewer: See if Gemini can create a Mermaid diagram for this based on the documentation.

Author: It didn't do a great job. I've left this as a TODO for now.

```mermaid
graph TD
    subgraph Client
        A[Client discovers AgentCard] --> B{Client activates extensions via X-A2A-Extensions header};
    end
    subgraph Agent
        C[AgentCard declares supported extensions] --> D{Agent identifies and activates supported extensions};
    end
    subgraph "Extension Specification"
        E[Defines extension URI and parameters];
    end
    B --> D;
    D -- Responds with activated extensions in X-A2A-Extensions header --> B;
    C -- Defines extension configuration --> E;
    A -- Reads AgentCard --> C
```

Another generated option:

```mermaid
graph TD
    subgraph "Definition & Discovery"
        A["Extension Specification Document <br>Defines URI, parameters, and logic"] -- "Defines" --> B;
        B["AgentCard <br>Declares supported extension via URI <br>Provides agent-specific configuration"] -- "Discovered by" --> D;
    end

    subgraph "Implementation"
        A -- "Guides" --> C{"Agent-Side Implementation <br>Built according to the specification"};
        A -- "Guides" --> D{"Client-Side Implementation <br>Built to support the extension"};
        C -- "Populates" --> B;
    end

    subgraph "Activation & Data Flow"
        D -- "Sends request with <br>X-A2A-Extensions header" --> E{"A2A Protocol <br>Data passes between Client and Agent"};
        E -- "Extension logic can change <br>the content of the data" --> E;
        C -- "Receives request and <br>activates extension logic" --> E;
    end
```

- Extensions are defined by a specification document
- The specification document defines the identifiers for the extension

> - The AgentCard declares support for an extension by referencing the identifier and providing any configuration

Reviewer: ```suggestion
- The `AgentCard` declares support for an extension by referencing the identifier and providing any configuration
```

- An implementation is built according to the specification
- A client has support for various extensions
- The client indicates extensions to activate in its request to the agent
- Extensions can change the content of data that passes over the protocol

## Extension Declaration

Agents declare their support for extensions in their `AgentCard` by including `AgentExtension` objects in their `AgentCapabilities` object.

- `uri`: The URI of the extension. This is an arbitrary identifier that the extension specification defines. Implementations of an extension use this URI to identify when to activate, and clients use this to determine extension compatibility.
- `required`: Whether the agent requires clients to use this extension.
- `description`: A description of how the agent uses the declared extension. Full details of a extension are intended to be in an extension specification. This field is useful to explain the connection between the agent and the extension.

> - `params`: Extension-specific configuration. The expected values to be placed in this field, if any, are defined by the extension specification. This field can be used for specifying parameters of the extension or declaring additional agent-specific data.

Reviewer: Try embedding the section of `types.ts` that has this information, or create a table similar to `specification.md`

[specification/#5-agent-discovery-the-agent-card](https://google-a2a.github.io/A2A/latest/specification/#5521-agentextension-object)

### Required Extensions

While extensions are a means of enabling additional functionality, we anticipate that some agents will have stricter requirements than those expressible by the core A2A protocol. For example, an agent may require that all incoming messages are cryptographically signed by their author. Extensions that are declared `required` are intended to support this use case.

> When an AgentCard declares a required extension, this is a signal to clients that some aspect of the extension impacts how requests are structured. Agents should not mark data-only extensions as required, since there is no direct impact on how requests are made to the agent.

Reviewer: ```suggestion
When an `AgentCard` declares a required extension, this is a signal to clients that some aspect of the extension impacts how requests are structured. Agents should not mark data-only extensions as required, since there is no direct impact on how requests are made to the agent.
```

> If an AgentCard declares a required extension, and the client does not request activation of that required extension, Agents should return reject the incoming request and return an appropriate error code.

Reviewer: ```suggestion
If an `AgentCard` declares a required extension, and the client does not request activation of that required extension, Agents should return reject the incoming request and return an appropriate error code.
```
